toolchain/check/member_access.cpp

```diff
@@ -366,4 +370,42 @@ auto PerformMemberAccess(Context& context, Parse::MemberAccessExprId node_id,
   return member_id;
 }
 
+auto PerformCompoundMemberAccess(Context& context,
+                                 Parse::MemberAccessExprId node_id,
+                                 SemIR::InstId base_id,
+                                 SemIR::InstId member_expr_id)
+    -> SemIR::InstId {
+  // Materialize a temporary for the base expression if necessary.
+  base_id = ConvertToValueOrRefExpr(context, base_id);
+  auto base_type_id = context.insts().Get(base_id).type_id();
+  auto base_type_const_id = context.types().GetConstantId(base_type_id);
+
+  auto member_id = member_expr_id;
+  auto member = context.insts().Get(member_id);
+
+  // If the member expression names an associated entity, impl lookup is always
+  // performed using the type of the base expression.
+  if (auto assoc_type = context.types().TryGetAs<SemIR::AssociatedEntityType>(
+          member.type_id())) {
+    member_id =
+        PerformImplLookup(context, base_type_const_id, *assoc_type, member_id);
+  }
+
+  // Perform instance binding if we found an instance member.
+  member_id = PerformInstanceBinding(context, node_id, base_id, member_id);
+
+  // If we didn't perform impl lookup or instance binding, that's an error
+  // because the base expression is not used for anything.
```

Comment:
Just flagging that there is a comment from me in the bindings proposal that suggests this might not always be an error.

Probably not worth changing anything at this stage, but flagging in review just in case you want to add a TODO or anything.